File: CPP/No75.cpp
/**
 * Created by Xiaozhong on 2020/7/24.
 * Copyright (c) 2020/7/24 Xiaozhong. All rights reserved.
 */

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    void sortColors(vector<int> &nums) {
        // 三指针法，left 指向 0， right 指向 2， cur 指向当前
        int n = nums.size(), left = 0, right = n - 1, cur = 0;
        while (cur <= right) {
            if (nums[cur] == 0) swap(nums[cur++], nums[left++]);
            else if (nums[cur] == 2) swap(nums[cur], nums[right--]);
            else cur++;
        }
    }

private:
    void swap(int &i, int &j) {
        int tmp = i;
        i = j;
        j = tmp;
    }
};

int main() {
    Solution s;
    vector<int> nums = {1, 2, 0};
    s.sortColors(nums);
    for (int i : nums) {
        cout << i << " ";
    }
}

File: Java/src/No75.java
/**
 * Java
 *
 * @Description: TODO
 * @Author: cherry
 * @Create on: 2022/8/9
 **/

public class No75 {
    public void sortColors(int[] nums) {
        int n = nums.length;
        int left = 0, right = n - 1;
        for (int idx = 0; idx <= right; ++idx) {
            while (idx <= right && nums[idx] == 2) swap(nums, idx, right--);
            if (nums[idx] == 0) swap(nums, left++, idx);
        }
    }

    private void swap(int[] nums, int l, int r) {
        int temp = nums[l];
        nums[l] = nums[r];
        nums[r] = temp;
    }

    public static void main(String[] args) {
        int[] nums = new int[]{2, 0, 2, 1, 0, 1, 1, 1, 0};
        new No75().sortColors(nums);
        System.out.println();
    }
}
